src/lib.rs

```diff
@@ -45,15 +45,74 @@ pub struct Header {
     pub generator: u32,
 }
 
+/// Forces early fragment tests. In a standard situation if a driver determines that it is possible to
+/// switch on early depth test it will. Typical situations when early depth test is switched off:
+///   - Calling ```discard``` in a shader.
+///   - Writing to the depth buffer, unless ConservativeDepthTest is enabled.
+///
+/// Enabling this option will do 2 things:
+///   - Disables any discard and depth write.
+///   - Ensures that image load/store operations will only happen on fragments that pass the depth test.
+///
+/// SPIR-V: ExecutionMode EarlyFragmentTests
+/// In GLSL: layout(early_fragment_tests) in;
+/// HLSL: Attribute earlydepthstencil
+///
+/// For more, see:
+///   - https://www.khronos.org/opengl/wiki/Early_Fragment_Test#Explicit_specification
+///   - https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/sm5-attributes-earlydepthstencil
+#[derive(Clone, Copy, Debug, PartialEq, Default)]
+#[cfg_attr(feature = "serialize", derive(Serialize))]
+#[cfg_attr(feature = "deserialize", derive(Deserialize))]
+pub struct ForcedEarlyDepthTest;
+
+/// States that this shader modifies depth of a fragment.
+///
+/// SPIR-V: ExecutionMode DepthReplacing
+#[derive(Clone, Copy, Debug, PartialEq)]
+#[cfg_attr(feature = "serialize", derive(Serialize))]
+#[cfg_attr(feature = "deserialize", derive(Deserialize))]
+pub struct FragmentDepthReplacing;
```

Comment:
I don't understand why this is needed in the IR. If we see that the entry point statically accesses `gl_FragDepth` then it's considered to be modifying the depth. No need for extra redundant execution modes.

Reply:
Validation: when outputting SPIR-V you must write this otherwise it fails Vulkan's module validation rules. Also, the other way around.

And you have to store it somewhere for the purpose above.

Reply:
Removed.